proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1417 @@
+     header (codomain). ``R`` starts empty.
+
+   * Let ``tv1 .. tvn`` represent the variables mentioned in the
+     declaration header for ``T``.
+
+   * Look at the result type of ``K`` and extract out the
+     type arguments to ``T``. Call these arguments ``arg1 .. argn``,
+     where ``T`` has arity ``n``.
+
+   * For each ``argi``: if ``argi`` is a bare variable ``v`` that is not
+     already included in the domain of ``R``, add a mapping to ``R``
+     from ``v`` to ``tvi``.
+
+   * The key set of ``R`` is precisely the set of universal variables
+     of ``K``; any other variable introduced in ``K``\ 's type is an
+     existential.
+
+   * During kind inference, treat an occurrence of a variable ``v``
+     in the key set of ``R`` as if it were an occurrence of ``R(v)``.
+
+   Note that the algorithm above does *not* depend on type or kind inference;
+   it is a straightforward pass over the abstract syntax of the constructor.
+
+   This version preserves today's scoping rules (and upholds the LLSP_). It
+   is backward-compatible. But it is subtler. Maybe it is better, regardless.
+
+#. We could offer users a migration period, where we warn about this
+   impending change. I see no easy way of implementing such a check,
+   and I see relatively little value in doing so, given that the fixes
+   are really quite easy. This opinion may change in the light of experience,
+   if this feature is implemented and we see trouble in the wild.
+
+.. _pattern-type-args:
+
+Type arguments in constructor patterns
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This is an update to accepted, implemented proposal `#126`_ that changes
+its treatment of universals. It incorporates the logic of not-yet-accepted
+amendment `#291`_.
+
+Specification
+^^^^^^^^^^^^^
+
+1. Introduce a new extension ``-XTypeAbstractions``.
+
+#. When ``-XTypeAbstractions`` is enabled, allow type application syntax
+   in constructor patterns.
+
+   Concretely, the grammar goes from ::
+
+     pat → gcon apat1 … apatk
+         …
+
+   to ::
+
+       pat → gcon tyapp_or_pat1 … tyapp_or_patk
+           …
+
+       tyapp_or_pat → '@' atype    -- '@' is in prefix position
+                    → pat
+
+#. Type applications in constructor patterns do *not* affect whether
+   the pattern-match is successful.
+
+#. Type applications in constructor patterns must correspond to ``forall … .``
+   quantifications in the declared constructor or pattern synonym type.
+   (Right now, pattern synonyms require all such quantifications to occur
+   before any term arguments, but accepted proposal `#402`_ allows these
+   quantifications to occur in any order in data constructors.
+
+#. Each quantification in a data constructor or pattern synonym brings
+   into scope either a universal variable or an existential variable.
+   Telling these apart is easy in pattern synonym types; `see above <#gadt-syntax>`_ for how to determine this property of data constructor
```

Comment:
a lot of these intra-document links don't seem to work for me (however, the table of contents, the link pointing to `gadt-alternatives`, and the links pointing to the principles do work.)
No idea why, though.